File: apps/ios/Sources/CrossyUI/RoomWeather.swift
// Honest weather (apps/ios/DESIGN.md §8): three connection states, three registers
// (PROTOCOL.md §7). Live is a calm dot; resyncing is a breathing dot and nothing
// else changes (the snapshot applies wholesale when it lands, the board keeps its
// last truth); reconnecting dims the room with a quiet countdown and names itself,
// because a person lost something mid-solve. Never a modal, never a spinner over the
// grid. The client-local `connecting` state (pre-first-welcome, no board truth yet) is
// the QUIET register: the dimmed dot beside the clock, no word and no countdown, since
// a first join has lost nothing (redesign 2026-07-11, the terse first-connect pill).
// The board still dims, mirroring the web's de-emphasized pre-welcome grid.
// The mapping is a pure function of the store's SyncState so tests pin it; the dot
// itself is achromatic (people are the only color, DESIGN.md §3).

import CrossyStore
import Foundation

/// What the room renders for one connection state: the dot's register on the
/// chrome, and whether the board dims. Countdown presence is reconnecting-only.
public struct RoomWeather: Equatable, Sendable {
    public enum Dot: Equatable, Sendable {
        /// A quiet, steady dot: the room is live.
        case calm
        /// The dot breathes (a slow opacity pulse): a gap was seen, a snapshot is
        /// on its way. Chrome-only; the board does not change.
        case breathing
        /// The dot holds hollow while the room is dimmed: the socket is gone.
        case dimmed
    }

    public let dot: Dot
    /// True dims the board under the chrome (reconnecting and the pre-welcome
    /// connecting state). The dim is a paper wash, never a modal or a spinner.
    public let boardDimmed: Bool
    /// True shows the quiet countdown to the next dial (reconnecting only).
    public let showsCountdown: Bool
    /// The plain word for the state, nil when the room needs no word (ID-5: common
    /// words, nothing precious).
    public let label: String?

    public init(dot: Dot, boardDimmed: Bool, showsCountdown: Bool, label: String?) {
        self.dot = dot
        self.boardDimmed = boardDimmed
        self.showsCountdown = showsCountdown
        self.label = label
    }

    /// The one mapping (PROTOCOL.md §7 states plus the client-local connecting).
    public static func from(sync: SyncState) -> RoomWeather {
        switch sync {
        case .live:
            return RoomWeather(dot: .calm, boardDimmed: false, showsCountdown: false, label: nil)
        case .resyncing:
            return RoomWeather(dot: .breathing, boardDimmed: false, showsCountdown: false, label: nil)
        case .reconnecting:
            return RoomWeather(
                dot: .dimmed, boardDimmed: true, showsCountdown: true, label: "Reconnecting")
        case .connecting:
            // The FIRST connect is quiet (DESIGN.md §8: never a spinner; the room's law
            // is a hush, not a status word). A reconnect exists because a person LOST
            // something mid-solve, so it names itself and counts down; a first join has
            // lost nothing, so the pill carries only the dimmed dot beside the clock, no
            // word and no countdown. The board still dims (the room is honestly not live
            // yet), mirroring the web's de-emphasized pre-welcome grid.
            return RoomWeather(
                dot: .dimmed, boardDimmed: true, showsCountdown: false, label: nil)
        }
    }

    /// The board's paper wash opacity when dimmed: strong enough to read as an
    /// honest hush, light enough that the room never dims dead.
    public static let boardDimOpacity: Double = 0.45

    /// The reconnect overlay grace (Track A-ios; twin of the web's
    /// RECONNECT_OVERLAY_GRACE_MS, 2000 ms). The non-live pair (resyncing,
    /// reconnecting; PROTOCOL.md §7) shows its weather only after the connection
    /// has been continuously non-live this long. Railway's edge recycles a
    /// healthy socket on a schedule and reconnect-and-resync heals in ~200 ms, so
    /// a bare recycle must never flash the overlay. Presentation only: the store's
    /// state machine, the transport, and the backoff are untouched.
    public static let reconnectOverlayGraceSeconds: Double = 2

    /// Whole seconds until the next dial, floored at zero; nil when there is no
    /// deadline to count toward (the adapter has not scheduled one).
    public static func countdownSeconds(retryAt: Date?, now: Date) -> Int? {
        guard let retryAt else { return nil }
        return max(0, Int(retryAt.timeIntervalSince(now).rounded(.up)))
    }

    /// The countdown line, ID-5 plain and warm: "Back in 3s" while a dial is
    /// scheduled, the bare state word otherwise.
    public static func reconnectLine(retryAt: Date?, now: Date) -> String {
        guard let seconds = countdownSeconds(retryAt: retryAt, now: now), seconds > 0 else {
            return "Reconnecting"
        }
        return "Back in \(seconds)s"
    }
}

/// The grace gate for the non-live pair (PROTOCOL.md §7 resyncing/reconnecting).
/// Pure and clock-injected, the RoomWeather.countdownSeconds posture: the view
/// model folds observations and reads a `now`, so tests pin the timing headlessly.
/// Live and connecting clear the origin, so recovery hides the overlay
/// immediately; the non-live pair shares ONE origin, so a bounce between
/// resyncing and reconnecting never restarts the grace and the overlay never
/// flickers. The first-connect `connecting` register keeps its own quiet handling
/// and is never gated here (a first join has lost nothing).
public struct ReconnectOverlayGate: Equatable, Sendable {
    /// The instant the connection last went non-live from a live or connecting
    /// state, nil while live or connecting.
    public private(set) var nonLiveSince: Date?

    public init() {}

    /// Fold one connection-state observation at `now`.
    public mutating func observe(_ sync: SyncState, now: Date) {
        switch sync {
        case .live, .connecting:
            nonLiveSince = nil
        case .resyncing, .reconnecting:
            if nonLiveSince == nil { nonLiveSince = now }
        }
    }

    /// True once the connection has been continuously non-live for the grace
    /// window. False while live or connecting, so recovery hides at once.
    public func overlayPresented(now: Date) -> Bool {
        guard let since = nonLiveSince else { return false }
        return now.timeIntervalSince(since) >= RoomWeather.reconnectOverlayGraceSeconds
    }

    /// The sync state the room presents: the non-live pair reads as `live` until
    /// the grace elapses (calm dot, board undimmed, input on), then reveals its
    /// true register; `live` and `connecting` pass through untouched.
    public func presentedSync(_ sync: SyncState, now: Date) -> SyncState {
        switch sync {
        case .live, .connecting:
            return sync
        case .resyncing, .reconnecting:
            return overlayPresented(now: now) ? sync : .live
        }
    }

    /// Seconds until the overlay would present, nil when it already shows or the
    /// connection is live or connecting (nothing to wake for). The view model arms
    /// a one-shot timer with this so the room re-renders into the overlay when the
    /// grace elapses with no further state change.
    public func secondsUntilPresented(now: Date) -> Double? {
        guard let since = nonLiveSince else { return nil }
        let remaining = RoomWeather.reconnectOverlayGraceSeconds - now.timeIntervalSince(since)
        return remaining > 0 ? remaining : nil
    }
}

File: apps/android/ui/src/main/kotlin/crossy/ui/RoomWeather.kt
// Honest weather (apps/ios/DESIGN.md §8; apps/ios/Sources/CrossyUI/RoomWeather.swift): the room's
// connection state as behavior, not chrome polish. Reconnecting dims the room with a quiet countdown
// because a person lost something mid-solve; the first connect (client-local `connecting`, no board
// truth yet) also dims but says nothing and counts nothing, since a first join has lost nothing. Live
// and resyncing leave the board alone (a snapshot applies wholesale when it lands). This is the pure
// mapping of the store's SyncState plus the driver's next-dial deadline; tests pin it. The dot's
// three registers (calm / breathing / dimmed) live here too now (the room bar draws them, RoomBar's
// WeatherDot); the glass chip is a later material track, so the register is the behavior, not polish.

package crossy.ui

import crossy.store.SyncState
import kotlin.math.ceil
import kotlin.math.max
import kotlinx.coroutines.ExperimentalCoroutinesApi
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.distinctUntilChanged
import kotlinx.coroutines.flow.flatMapLatest
import kotlinx.coroutines.flow.flow
import kotlinx.coroutines.flow.flowOf
import kotlinx.coroutines.flow.map

object RoomWeather {
    /** The grace window a lost connection waits out before the room announces it (the web and iOS
     *  twins carry the same 2000 ms). Railway's edge proxy cycles drop healthy sockets, and a
     *  reconnect-and-resync measures ~200 ms end to end, so the raw non-live state would flicker the
     *  weather pill and dim the board for a blink. Chrome waits this long before it speaks. */
    const val RECONNECT_OVERLAY_GRACE_MS: Long = 2000L

    /** The debounced flag every reconnect/resync surface reads (the pill's word and dot, the board
     *  dim): true once [sync] has stood continuously non-live for RECONNECT_OVERLAY_GRACE_MS, false
     *  the instant it returns to live. One shared timer covers both non-live registers, so a
     *  resyncing<->reconnecting bounce reads as one unbroken non-live stretch and cannot flash the
     *  chrome; flatMapLatest cancels the pending timer on recovery, so a live snapshot clears the
     *  overlay at once. Presentation only: input is never gated by this (the grid and fan stay live
     *  through the window), it only holds back the announcement. */
    @OptIn(ExperimentalCoroutinesApi::class)
    fun overlayGrace(sync: Flow<SyncState>): Flow<Boolean> =
        sync.map { it != SyncState.LIVE }
            .distinctUntilChanged()
            .flatMapLatest { nonLive ->
                if (nonLive) flow { delay(RECONNECT_OVERLAY_GRACE_MS); emit(true) } else flowOf(false)
            }
            .distinctUntilChanged()

    /** The weather dot's three registers (apps/ios RoomWeather.Dot, DESIGN.md §8): a calm steady dot
     *  when live, a breathing pulse while a snapshot is on its way (resyncing), and a hollow ring
     *  while the room is dimmed (connecting or reconnecting). Achromatic: weather is the room's state,
     *  not a person (DESIGN.md §3). */
    enum class Dot {
        /** A quiet, steady dot: the room is live. */
        CALM,

        /** The dot breathes (a slow opacity pulse): a gap was seen, a snapshot is on its way.
         *  Chrome-only; the board does not change. */
        BREATHING,

        /** The dot holds hollow while the room is dimmed: the socket is gone, or not yet arrived. */
        DIMMED,
    }

    /** The dot register for one connection state (PROTOCOL.md §7 states plus the client-local
     *  connecting): live is calm, resyncing breathes, and both the pre-welcome connecting and a
     *  post-drop reconnecting hold the hollow dimmed dot. Twin of iOS RoomWeather.from(sync).dot. */
    fun dot(sync: SyncState): Dot = when (sync) {
        SyncState.LIVE -> Dot.CALM
        SyncState.RESYNCING -> Dot.BREATHING
        SyncState.CONNECTING, SyncState.RECONNECTING -> Dot.DIMMED
    }

    /** The plain word beside the dot, or null when the room needs none (ID-5: common words, nothing
     *  precious). Only a reconnect names itself, because a person lost something mid-solve; live,
     *  resyncing, and the first connect stay wordless (the terse first-connect pill, iOS
     *  RoomWeather.from). The reconnect countdown replaces this word while a dial is scheduled
     *  (showsCountdown + reconnectLine). Twin of iOS RoomWeather.label. */
    fun label(sync: SyncState): String? =
        if (sync == SyncState.RECONNECTING) "Reconnecting" else null

    /** The board's paper wash opacity when dimmed: strong enough to read as an honest hush, light
     *  enough that the room never dims dead (RoomWeather.boardDimOpacity). */
    const val boardDimOpacity: Double = 0.45

    /** True dims the board under the chrome: the pre-welcome `connecting` state and a post-drop
     *  `reconnecting`, both honestly not live. Live and resyncing keep the board's last truth. */
    fun boardDimmed(sync: SyncState): Boolean =
        sync == SyncState.CONNECTING || sync == SyncState.RECONNECTING

    /** True shows the quiet countdown to the next dial: reconnecting only (a first connect has lost
     *  nothing, so it names nothing). */
    fun showsCountdown(sync: SyncState): Boolean = sync == SyncState.RECONNECTING

    /** Whole seconds until the next dial, floored at zero; null when there is no deadline to count
     *  toward (the driver has not scheduled one). Epoch millis on Android, the SessionDriver clock. */
    fun countdownSeconds(retryAtMillis: Long?, nowMillis: Long): Int? {
        if (retryAtMillis == null) return null
        return max(0.0, ceil((retryAtMillis - nowMillis) / 1000.0)).toInt()
    }

    /** The countdown line, ID-5 plain and warm: "Back in 3s" while a dial is scheduled, the bare
     *  state word otherwise (no deadline yet, or it has already elapsed). Twin of reconnectLine. */
    fun reconnectLine(retryAtMillis: Long?, nowMillis: Long): String {
        val seconds = countdownSeconds(retryAtMillis, nowMillis)
        return if (seconds != null && seconds > 0) "Back in ${seconds}s" else "Reconnecting"
    }
}
